File: ExcelRibbon/LinksAnalyzer/LinksLexer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ExcelRibbon.LinksAnalyzer {
    public enum Token { 
        ScanError,
        EOT,
        Identifier,
        StringLiteral,
        Number,
        BinOp,
        Unop,
        Equals,
        Comma,
        SemiColon,
        Bang,
        OpenParen,
        CloseParen,
        ExternRef,
        OpenBrace,
        CloseBrace
    }

    public interface IToken {
        Token   Token   { get; }
        string  Text    { get; }
        long    Start   { get;}
    }

    internal static partial class Extensions {
        public static IToken Set(this Token token, int start, LinksLexer lexer)
            => lexer.Set(token, start, lexer.CharPosition);

        public static bool IsNumeric (this char c) => '0' <= c && c <= '9';
        public static bool IsAlpha (this char c) => 'A' <= c && c <= 'Z';
        public static bool IsAlphanumeric(this char c) => c.IsAlpha() || c.IsNumeric();
        public static bool IsWhiteSpace(this char c) => c == ' ' || c == '\n' || c == '\r' || c == '\t';
    }

    public interface ILinksLexer {

    }

    public class LinksLexer {
        public LinksLexer(InternalCellRef cellRef, string textIn) {
            ParenDepth = BraceDepth = 0;
            CellRef      = cellRef;
            TextIn       = textIn;
            CharPosition = 0;
        }

        private static IReadOnlyList<string> WordOperators = new List<string> { "AND", "OR" };

        private char CurrentCharacter;

        private InternalCellRef CellRef { get; set; }
        public  string TextIn       { get; }
        public  int    CharPosition { get; private set; }
        public  bool   IsEOT        => CharPosition >= TextIn.Length;
        private int    ParenDepth   { get; set; }
        private int    BraceDepth   { get; set; }
        private string TokenText    { get; set; }

        private string  Formula => TextIn;

        private ParseError ErrorHere(string source, string condition) =>
            new ParseError(CellRef, Formula, CharPosition, $"{source}: {condition}");

        internal IToken Scan() {
            while ( Advancable() ) {
                if ( ! IsWhiteSpace ) {
                    var start = CharPosition;
                    switch (CurrentCharacter) {
                        case '!': return Token.Bang.Set(start, this);
                        case '=': return Token.Equals.Set(start, this);
                        case ',': return Token.Comma.Set(start, this);
                        case ';': return Token.SemiColon.Set(start, this);
                        case '+':
                        case '-':
                        case '%': return Token.Unop.Set(start, this);
                        case '*':
                        case '/':
                        case '&':
                        case '^': return Token.BinOp.Set(start, this);
                        case '<':
                        case '>': if ( IsEOT ) { break; }
                                  if (NextCharacterIs('=')) { CharPosition++; }
                                  return Token.BinOp.Set(start, this);
                        case '(': ParenDepth++; return Token.OpenParen.Set(start, this);
                        case ')': ParenDepth--; return Token.CloseParen.Set(start, this);
                        case '{': BraceDepth++; return Token.OpenBrace.Set(start, this);
                        case '}': BraceDepth--; return Token.CloseBrace.Set(start, this);
                        case '#': return ScanErrorIdent(start);
                        case '"': return ScanStringLiteral(start);
                        case '\'': return ScanExternalRef(start);
                        default:
                            if (CurrentCharacter.IsNumeric()) return ScanNumber(start);
                            if ( CurrentCharacter.IsAlpha() ) return ScanIdent(start);
                            return Set(Token.ScanError, start, CharPosition);
                    }
                }
            }
            return Token.EOT.Set(CharPosition, this);
        }

        private IToken ScanExternalRef(int start) {
            while ( ! Advancable() ) {
                if (CurrentCharacter == '\'') {
                    if (Advancable() ) { 
                        return Token.ScanError.Set(start, this);
                    } else  if (CurrentCharacter != '\'') {
                        return Token.ExternRef.Set(start, this);
                    }
                }
            }
            return Token.ExternRef.Set(start, this);
        }


        private IToken ScanStringLiteral(int start) {
            while ( Advancable() ) {
                if (CurrentCharacter == '"') {
                    if (! Advancable()  ||  IsWhiteSpace) {
                        return Token.StringLiteral.Set(start, this);
                    } else if (CurrentCharacter == '"') {
                        continue;
                    } else {
                        return Token.StringLiteral.Set(start, this);
                    }
                }
            }
            return Set(Token.ScanError, start, CharPosition);
        }

        internal IToken ScanIdent(int start) {
            while ( Advancable()) {
                if (IsWhiteSpace) { return Token.Identifier.Set(start, this); }
                if (CurrentCharacter.IsAlphanumeric() 
                ||  CurrentCharacter == '_'
                ||  CurrentCharacter == ':') {
                    continue;
                } else {
                    return Token.Identifier.Set(start, this);
                }
            }
            return Token.Identifier.Set(start, this);
        }

        internal IToken ScanErrorIdent(int start) => ScanIdent(start);

        internal IToken ScanNumber(int start) {
            bool ParsingFraction = false;
            while (Advancable()) {
                if (CurrentCharacter.IsNumeric()) {
                    continue;
                } else if (CurrentCharacter != '.') {
                    return Token.Number.Set(start, this);
                } else if (ParsingFraction) {
                    return Token.ScanError.Set(start, this);
                } else {
                    ParsingFraction = true;
                }
            }
            return Token.Number.Set(start, this);
        }

        internal IToken Set(Token token, int start, int end) =>
            new TokenX(token, start, TextIn.Substring(start, end - start));

        /// <summary>If not EOT scans forward snd returns true if that charaacter is white space..</summary>
        /// <returns></returns>
        private bool Advancable() {
            if ( IsEOT ) { 
                return false;
            } else {
                CurrentCharacter = TextIn[CharPosition++];
                return true;
            }
        }

        private bool NextCharacterIs(char c) => ! IsEOT && TextIn[CharPosition + 1] == c;

        private bool IsWhiteSpace => 
            CurrentCharacter == '\r' || CurrentCharacter == '\n' ||
            CurrentCharacter == '\t' || CurrentCharacter == ' ';

        internal class TokenX : IToken {
            public TokenX(Token token, int start, string text) {
                Token = token;
                Start = start;
                Text  = text;
            }

            public Token    Token   { get; }
            public long     Start   { get;}
            public string   Text    { get; }
        }

        private bool IsWordOperator(string text) => WordOperators.FirstOrDefault(s => s == text) != null;

        internal ParseError VerifyNextToken(Token ExpectedType, string ExpectedText) {
            if (Scan(TokenText).Token == ExpectedType && TokenText == ExpectedText) return null;
            return ErrorHere("VerifyNextToken", $"Expected: '{ExpectedText}'");
        }

        public ParseError VerifyBraceDepth(int expected) {
            if (BraceDepth == expected) return null;
            return ErrorHere("VerifyBraceDepth", $" depth = {BraceDepth}; expected {expected}.");
        }
        public ParseError VerifyParenDepth(int expected) {
            if (ParenDepth == expected) return null;
            return ErrorHere("VerifyParenDepth", $" depth = {ParenDepth}; expected {expected}.");
        }

        public ParseError RaiseError(InternalCellRef cellRef, string expectedText) =>
            new ParseError(cellRef, TextIn, CharPosition, $"{expectedText} at position {CharPosition}");
    }
}

File: LinksAnalyzerVBA/LinksLexer.vb
Public Class LinksLexer
#Const ShowError = 0

    Private Const mModuleName As String = "Lexer."

    Private mIsInitialized As Boolean
    Private mTextIn As String
    Private mCharPos As Long
    Private mCurrentChar As String

    Private mWordOperators As Collection

    Private mBraceDepth As Long
    Private mParenDepth As Long

    Public Property Get Text() As String
    Text = mTextIn
End Property
    Public Function LoadText(ByVal TextIn As String) As LinksLexer
        Const MethodName As String = mModuleName & "Path"

        On Error GoTo EH
        mTextIn = TextIn
        mCharPos = 0
        Advance()
            
    Set LoadText = Me
    mIsInitialized = True
XT:     Exit Function

EH:     ReRaiseError Err, MethodName
End Function

    Friend Function Scan(ByRef TokenText As String) As Token
        Const MethodName As String = mModuleName & "Scan"
        ErrorUtils.CheckIsInitialized mIsInitialized, MethodName

    On Error GoTo EH
        Do
            If IsEOT Then Scan = EOT : GoTo XT
            Select Case mCurrentChar
                Case vbCr, vbLf, vbTab, " " : Advance()
                Case Else : GoTo NoMoreWhiteSpace
            End Select
        Loop

NoMoreWhiteSpace:
        Dim TokenStart As Long : TokenStart = mCharPos
        Select Case VBA.UCase(mCurrentChar)
            Case "!" : Advance() :
                Scan = Bang
            Case "=" : Advance() :
                Scan = Equals()
            Case "," : Advance() :
                Scan = Comma
            Case ";" : Advance() :
                Scan = SemiColon
            Case "+", "-", "%" : Advance() :
                Scan = Unop
            Case "*", "/", "&", "^", "<", ">" : Advance() :
                Scan = BinOp
            Case "<", ">" : Advance() :
                If mCurrentChar = "=" Then Advance() :
                Scan = BinOp
            Case "(" : Advance() :
                Scan = OpenParen : mParenDepth = mParenDepth + 1
            Case ")" : Advance() :
                Scan = CloseParen : mParenDepth = mParenDepth - 1
            Case "{" : Advance() :
                Scan = OpenBrace : mBraceDepth = mBraceDepth + 1
            Case "}" : Advance() :
                Scan = CloseBrace : mBraceDepth = mBraceDepth - 1
            Case "A" To "Z", "_", "$" : Scan = ScanIdent()
            Case "#" : Scan = ScanErrorIdent()
            Case "0" To "9" : Scan = ScanNumber()
            Case "'" : Scan = ScanExternRef()
            Case """" : Scan = ScanStringLiteral()
            Case Else : Advance() :
                Scan = ScanError
        End Select

        TokenText = VBA.mID$(mTextIn, TokenStart, mCharPos - TokenStart)

        If Scan = Identifier Then If IsWordOperator(TokenText) Then Scan = BinOp
XT:     Exit Function

EH:     ReRaiseError Err, MethodName
End Function

    Friend Sub VerifyNextToken(ByVal ExpectedType As Token, ByVal ExpectedText As String)
        Const MethodName As String = mModuleName & "VerifyNextToken"
        Dim TokenText As String
        If Scan(TokenText) <> ExpectedType Or TokenText <> ExpectedText Then _
            Err.Raise 1, MethodName, "Expected: '" & ExpectedText & "'"
End Sub

    Public Function VerifyParenDepth(ByVal ExpectedDepth As Long)
        Const MethodName As String = mModuleName & "VerifyParenDepthExpected"
        If mParenDepth <> ExpectedDepth Then _
        Err.Raise 1, MethodName, "Paren depth = " & mParenDepth & "; expected " & ExpectedDepth
End Function

    Public Function VerifyBraceDepth(ByVal ExpectedDepth As Long)
        Const MethodName As String = mModuleName & "VerifyParenDepthExpected"
        If mParenDepth <> ExpectedDepth Then _
        Err.Raise 1, MethodName, "Brace depth = " & mBraceDepth & "; expected " & ExpectedDepth
End Function

    Private Function IsWordOperator(ByVal Text As String) As Boolean
        Const MethodName As String = mModuleName & "Scan"

        On Error GoTo EH
        IsWordOperator = False
        IsWordOperator = mWordOperators.Item(Text) <> ""
XT:     Exit Function

EH:     If Err.Number = 5 Then Resume Next  ' Subscript out of range => Not Found
        ReRaiseError Err, MethodName
End Function

    Friend Function ScanExternRef() As Token
        Const MethodName As String = mModuleName & "ScanExternRef"
        ErrorUtils.CheckIsInitialized mIsInitialized, MethodName

    On Error GoTo EH
        ScanExternRef = ScanError
        Do
            If Advance() Then GoTo XT
            If mCurrentChar = "'" Then
                If Advance() Then GoTo XT
                If mCurrentChar <> "'" Then GoTo ExitLoop
            End If
        Loop
ExitLoop:
        ScanExternRef = ExternRef

XT:     Exit Function

EH:     ReRaiseError Err, MethodName
End Function

    Friend Function ScanStringLiteral() As Token
        Const MethodName As String = mModuleName & "ScanStringLiteral"
        ErrorUtils.CheckIsInitialized mIsInitialized, MethodName

    On Error GoTo EH
        ScanStringLiteral = StringLiteral
        Do While Not Advance()

            If mCurrentChar = """" Then
                If Advance() Then GoTo XT
                If mCurrentChar <> """" Then GoTo XT
                If Advance() Then GoTo XT
            End If
        Loop
        ScanStringLiteral = ScanError

XT:     Exit Function

EH:     ReRaiseError Err, MethodName
End Function

    Friend Function ScanIdent() As Token
        Const MethodName As String = mModuleName & "ScanIdent"
        ErrorUtils.CheckIsInitialized mIsInitialized, MethodName

    On Error GoTo EH
        ScanIdent = ScanError
        Do
            Select Case mCurrentChar
                Case "A" To "Z", "_", "$", ":", "0" To "9" ' NO-OP
                Case Else : Exit Do
            End Select
        Loop Until Advance()

        ScanIdent = Identifier
XT:     Exit Function

EH:     ReRaiseError Err, MethodName
End Function

    Friend Function ScanErrorIdent() As Token
        Const MethodName As String = mModuleName & "ScanErrorIdent"
        ErrorUtils.CheckIsInitialized mIsInitialized, MethodName

    On Error GoTo EH
        Advance()
        If ScanIdent() <> Identifier Then GoTo ScanError

        Select Case mCurrentChar
#If ShowError = 1 Then
    Case "!":       Advance
#Else
            Case "!", "?" : Advance()
#End If
                ScanErrorIdent = Identifier
            Case "/" : Advance()
                If ScanIdent() <> Identifier Then GoTo ScanError
                ScanErrorIdent = Identifier
            Case Else : GoTo ScanError
        End Select

XT:     Exit Function

ScanError:
        ScanErrorIdent = ScanError
        GoTo XT

EH:     ReRaiseError Err, MethodName
End Function

    Friend Function ScanNumber() As Token
        Const MethodName As String = mModuleName & "ScanNumber"
        ErrorUtils.CheckIsInitialized mIsInitialized, MethodName

    On Error GoTo EH
        ScanNumber = ScanError
        Dim ParsingFraction As Boolean
        Do
            Select Case mCurrentChar
                Case "0" To "9"                ' NO-OP
                Case "."
                    If ParsingFraction Then ScanNumber = ScanError
                    ParsingFraction = True
                Case Else : Exit Do
            End Select
        Loop Until Advance()

        ScanNumber = Number
XT:     Exit Function

EH:     ReRaiseError Err, MethodName
End Function

    Public Property Get CharPos() As Long
    ErrorUtils.CheckIsInitialized mIsInitialized, mModuleName & "CharPos"
    CharPos = mCharPos
End Property

    ''' <summary>Returns true exactly when the last token on the string has been scanned.</summary>
    Public Property Get IsEOT() As Boolean
    IsEOT = mCharPos > Len(mTextIn)
End Property
    Public Property Get This() As LinksLexer
    ErrorUtils.CheckIsInitialized mIsInitialized, mModuleName & "CharPos"
    Set This = Me
End Property

    Public Function RaiseError(
    ByVal CellRef As InternalCellRef,
    ByVal ExpectedText As String
) As ParseError
        With New ParseError
        Set RaiseError = .Initialize(CellRef, mTextIn, CharPos, _
                ExpectedText & " at position " & CharPos)
    End With
    End Function

    ''' <summary>Advances current character by one space, and returns the value of IsEOT.</summary>
    ''' <returns>Returns true exactly when the last token on the string has been scanned.</returns>
    Private Function Advance() As Boolean
        If Not IsEOT Then
            mCharPos = mCharPos + 1
            If Not IsEOT Then mCurrentChar = VBA.UCase$(VBA.mID$(mTextIn, mCharPos, 1))
        End If
        Advance = IsEOT
    End Function

    Private Sub Class_Initialize()
        mIsInitialized = False
    Set mWordOperators = New Collection
    mWordOperators.Add "AND", "AND"
    mWordOperators.Add "OR", "OR"
End Sub

End Class
